src/platform/datapath_raw_socket.c

```diff
@@ -371,28 +370,72 @@ CxPlatResolveRoute(
     //
     // Map the next hop IP address to a link-layer address.
     //
-    MIB_IPNET_ROW2 IpnetRow = {0};
-    IpnetRow.InterfaceLuid = IpforwardRow.InterfaceLuid;
+    IpnetRow->InterfaceLuid = IpforwardRow.InterfaceLuid;
     if (QuicAddrIsWildCard(&IpforwardRow.NextHop)) { // On-link?
-        IpnetRow.Address = Route->RemoteAddress;
+        IpnetRow->Address = Route->RemoteAddress;
     } else {
-        IpnetRow.Address = IpforwardRow.NextHop;
+        IpnetRow->Address = IpforwardRow.NextHop;
     }
 
     //
-    // First call GetIpNetEntry2 to see if there's already a cached neighbor. If there
-    // isn't one, or if the cached neighbor's state is unreachable (which, NB, can happen
-    // in the case where a route lookup resulted in a dummy neighbor entry being created
-    // in TCPIP.sys) or incomplete, then queue up a solicitation event.
+    // Call GetIpNetEntry2 to see if there's already a cached neighbor.
     //
-    Status = GetIpNetEntry2(&IpnetRow);
+    Status = GetIpNetEntry2(IpnetRow);
+    if (Status != ERROR_SUCCESS || IpnetRow->State <= NlnsIncomplete) {
+        Status = ERROR_IO_PENDING;
+    } else {
+        CxPlatCopyMemory(
+            &Route->NextHopLinkLayerAddress,
+            IpnetRow->PhysicalAddress,
+            sizeof(Route->NextHopLinkLayerAddress));
+    }
+
+Done:
+    if (Status > 0) {
+        return SUCCESS_HRESULT_FROM_WIN32(Status);
+    } else {
+        return HRESULT_FROM_WIN32(Status);
+    }
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatResolveRoute(
+    _In_ CXPLAT_SOCKET* Socket,
+    _Inout_ CXPLAT_ROUTE* Route,
+    _In_ uint8_t PathId,
+    _In_ void* Context,
+    _In_ CXPLAT_ROUTE_RESOLUTION_CALLBACK_HANDLER Callback
+    )
+{
+#ifdef _WIN32
+    QUIC_STATUS Status = 0;
+    CXPLAT_ROUTE RouteQueried = *Route;
+    CXPLAT_ROUTE_STATE State = Route->State;
+    MIB_IPNET_ROW2 IpnetRow = {0};
+
     QuicTraceLogConnInfo(
         RouteResolutionStart,
         Context,
-        "Starting to look up neighbor on Path[%hhu] with status %u",
+        "Starting to look up neighbor on Path[%hhu] with status %d",
         PathId,
         Status);
-    if (Status != ERROR_SUCCESS || IpnetRow.State <= NlnsIncomplete) {
+
+    if (State == RouteSuspected) {
+        //
+        // We have seen hints that the route might be bad. Let the route worker take care of this.
+        //
+        Status = QUIC_STATUS_PENDING;
```

Comment:
Why don't you call CxPlatQueryRoute in this case? It might be that the new neighbor has already been resolved by some other connection, in which case we could do everything inline, right?

Reply:
We could but we'd have this "check if it's the same logic" in two places.

Reply:
I'm not sure what you mean. Is there a typo in your response?

Reply:
My point here is that "refreshing" and "resolving" the route should be nearly identical codepaths.

Reply:
=> We could but we'd need to add "check if it's the same neighbor" logic in two places.

Reply:
Nearly identical but still different. What are you suggesting here? Just do the same thing for both refresh and initial resolution?

Reply:
Yes, unless there's a strong reason not to. I mean think about it. We have basically the same goal in both cases: look up a new next hop MAC address. So why should the code paths be so divergent?